pyop2/types/dat.py

```diff
@@ -1021,6 +1021,8 @@ def vec_context(self, access):
                     size = v.local_size
                     array[offset:offset+size] = v.array_r[:]
                     offset += size
+
+        self._vec.stateIncrease()
```

Comment:
This should not be necessary because `self._vec.array` calls `VecGetArray/VecRestoreArray` which updates the state counter.

Reply:
It seems like something is going wrong then because:
```
f = Function(V*V)
f.dat._vec.stateGet() => 1
f.dat._vec.array = 2
f.dat._vec.stateGet() => 2
f.dat._vec.array[1:3] = 3
f.dat._vec.stateGet() => 3
```

Reply:
Ok thanks I will investigate that. Is the update of the state counter when `Vec._array` is called something recent ?

Increasing the state counter of the vec inside the `vec_context` is something that was originally in master so I assume there were a reason for that, e.g. handling a corner case.

Reply:
Oh I just realized that we don't call `Vec._array` in WRITE mode so it won't increase the state for that. That doesn't answer the question about why `Vec._array` is not increasing the counter inside the `vec_context` but probably explains why the state counter was originally increased with `self._vec.stateIncrease()`.

Reply:
> Oh I just realized that we don't call `Vec._array` in WRITE mode so it won't increase the state for that.

That is correct, because `access == WRITE` is a copy-out _from_ the single Vec _to_ the pieces (so it doesn't modify the single Vec.

Reply:
Ok then I guess that means there were a problem with the initial code as well as the increase was also done for `access == WRITE`. Also, I have got the following mfe:

```
f = Function(V * V)
f.dat._vec.stateGet() => 1
f.dat._vec.array[1:3] = 1
f.dat._vec.stateGet() => 2
array = f.dat._vec.array
array[1:3] = 2
f.dat._vec.stateGet() => 2
```

I am probably missing something obvious here but I would have expected that updating `array` updates the counter.

Reply:
Ah ok, there's some magic going on, `f.dat._vec.array` interfaces with the python buffer protocol, when that buffer is obtained, it calls `VecGetArray` and when the buffer is released it calls `VecRestoreArray`. It is the latter that increments the state counter. But in your code snippet you have not yet released the buffer (you're still holding a reference to `array`); if you put `del array` before the final `stateGet` then I think things should work.

Reply:
Ah ok thanks, it's clearer now. It indeed works and adding this to the `vec_context` fixes the failing tests. It is still unclear to me why the counter was initially incremented via `self._vec.stateIncrease()` but at least we have a clear solution now.